src/hotspot/share/gc/shenandoah/shenandoahOldGeneration.cpp

```diff
@@ -383,67 +374,68 @@ void ShenandoahOldGeneration::transition_to(State new_state) {
+//               |   |            v
+//               |   |          +-----------------+
+//               |   +--------> |     WAITING     |
+//               |   |          |  FOR BOOTSTRAP  |
+//               |   |          +-----------------+
+//               |   |            |
+//               |   |            | Reset Bitmap
+//               |   |            v
+//               |   |          +-----------------+     +----------------------+
+//               |   |          |    BOOTSTRAP    | <-> |       YOUNG GC       |
+//               |   |          |                 |     | (RSet Parses Region) |
+//               |   |          +-----------------+     +----------------------+
+//               |   |            |
+//               |   |            | Old Marking
+//               |   |            v
+//               |   |          +-----------------+     +----------------------+
+//               |   |          |     MARKING     | <-> |       YOUNG GC       |
+//               |   +--------- |                 |     | (RSet Parses Region) |
+//               |              +-----------------+     +----------------------+
+//               |                |
+//               |                | Has Evacuation Candidates
+//               |                v
+//               |              +-----------------+     +--------------------+
+//               |              |    EVACUATING   | <-> |      YOUNG GC      |
+//               +------------- |                 |     | (RSet Uses Bitmap) |
+//                              +-----------------+     +--------------------+
+//
+//
 //
 void ShenandoahOldGeneration::validate_transition(State new_state) {
   ShenandoahHeap* heap = ShenandoahHeap::heap();
   switch (new_state) {
-    case IDLE:
-      // GC cancellation can send us back to IDLE from any state.
-      assert(!heap->is_concurrent_old_mark_in_progress(), "Cannot become idle during old mark.");
-      assert(_old_heuristics->unprocessed_old_collection_candidates() == 0, "Cannot become idle with collection candidates");
-      assert(!heap->is_prepare_for_old_mark_in_progress(), "Cannot become idle while making old generation parseable.");
-      assert(heap->young_generation()->old_gen_task_queues() == nullptr, "Cannot become idle when setup for bootstrapping.");
-      break;
     case FILLING:
-      assert(_state == IDLE || _state == WAITING_FOR_FILL, "Cannot begin filling without first completing evacuations, state is '%s'", state_name(_state));
-      assert(heap->is_prepare_for_old_mark_in_progress(), "Should be preparing for old mark now.");
+      assert(heap->is_old_bitmap_stable(), "Cannot begin filling without first completing marking, state is '%s'", state_name(_state));
+      assert(_old_heuristics->has_coalesce_and_fill_candidates(), "Cannot begin filling without something to fill.");
+      break;
+    case WAITING_FOR_BOOTSTRAP:
+      // GC cancellation can send us back to IDLE from any state.
```

Comment:
Since there isn't now a state named "IDLE", it might be less confusing to just say "FILLING" or "WAITING_FOR_BOOTSTRAP"? Should we assert here that there are no `coalesce_and_fill_candidates` or is that implied by perhaps stronger checks below? (Wouldn't hurt even if so.)

Reply:
```
assert(_state != BOOTSTRAP && _state != WAITING_FOR_BOOTSTRAP, "Illegal transition");
```

Reply:
I am thinking that the most parsimonious way of covering all bases is by associating a pre-condition with entering each state and a post-condition with exiting each state, and verifying those conditions in the transition validation methods. But may be that can be done as a separate PR if it makes sense.

Reply:
Hmm - I thought that's what this code is doing. Or rather, it is asserting what conditions must hold to _be_ in this state (checking pre/post conditions simultaneously).